golang/Chapter05/model/client.go

```diff
@@ -0,0 +1,440 @@
+					pipeliner.Set(startKey, hourStart, 0)
+				} else if existing == 0 {
+					pipeliner.Set(startKey, hourStart, 0)
+				}
+
+				pipeliner.ZIncrBy(destination, 1, message)
+				c.LogRecent(name, message, severity, pipeliner)
+				return nil
+			}); err != nil {
+				log.Println("LogCommon pipelined failed, err: ", err)
+				return err
+			}
+			return nil
+		}, startKey)
+		if err != nil {
+			log.Println("watch failed, err: ", err)
+			continue
+		}
+	}
+}
+
+func (c *Client) UpdateCounter(name string, count int64, now int64) {
+	if now == 0 {
+		now = time.Now().Unix()
+	}
+
+	pipe := c.Conn.Pipeline()
+	for _, prec := range common.PRECISION {
+		pnow := (now / prec) * prec
+		hash := fmt.Sprintf("%d:%s", prec, name)
+		pipe.ZAdd("known:", &redis.Z{Member: hash, Score: 0})
+		pipe.HIncrBy("count:"+hash, strconv.Itoa(int(pnow)), count)
+	}
+	if _, err := pipe.Exec(); err != nil {
+		log.Println("updateCounter err: ", err)
+	}
+}
+
+func (c *Client) GetCount(name, precision string) [][]int {
+	hash := fmt.Sprintf("%v:%s", precision, name)
+	data := c.Conn.HGetAll("count:" + hash).Val()
+	toReturn := make([][]int, 0, len(data))
+	for k, v := range data {
+		temp := make([]int, 2)
+		key, _ := strconv.Atoi(k)
+		num, _ := strconv.Atoi(v)
+		temp[0], temp[1] = key, num
+		toReturn = append(toReturn, temp)
+	}
+	sort.Slice(toReturn, func(i, j int) bool {
+		return toReturn[i][0] < toReturn[j][0]
+	})
+	return toReturn
+}
+
+func (c *Client) CleanCounters() {
+	passes := 0
+
+	for !common.QUIT {
+		start := time.Now().Unix()
+		var index int64 = 0
+		for index < c.Conn.ZCard("known:").Val() {
+			hash := c.Conn.ZRange("known:", index, index).Val()
+			index++
+			if len(hash) == 0 {
+				break
+			}
+
+			hashValue := hash[0]
+			prec, _ := strconv.Atoi(strings.Split(hashValue, ":")[0])
+			bprec := prec / 60
+			if bprec == 0 {
+				bprec = 1
+			}
+			if passes%bprec != 0 {
+				continue
+			}
+
+			hkey := "count:" + hashValue
+			cutoff := int(time.Now().Unix()) - common.SAMPLECOUNT*prec
+			samples := c.Conn.HKeys(hkey).Val()
+			sort.Slice(samples, func(i, j int) bool {
+				return samples[i] < samples[j]
+			})
+			remove := sort.SearchStrings(samples, strconv.Itoa(cutoff))
+
+			if remove != 0 {
+				c.Conn.HDel(hkey, samples[:remove]...)
+				if remove == len(samples) {
+					err := c.Conn.Watch(func(tx *redis.Tx) error {
+						if tx.HLen(hkey).Val() == 0 {
+							pipe := tx.Pipeline()
+							pipe.ZRem("known:", hashValue)
+							_, err := pipe.Exec()
+							if err != nil {
+								log.Println("pipeline failed in CleanCounters: ", err)
+								return err
+							}
+							index--
+						} else {
+							tx.Unwatch()
+						}
+						return nil
+					}, hkey)
+
+					if err != nil {
+						log.Println("watch err in CleanCounters: ", err)
+						continue
+					}
+				}
+			}
+		}
+		passes++
+		duration := utils.Min(time.Now().Unix()-start+1, 60)
+		time.Sleep(time.Duration(utils.Max(60-duration, 1)) * time.Minute)
+	}
+}
```

Comment:
Since I'm in here and reading anyway, I feel really bad about this counter in Redis. There are so many different / better ones I've written personally, never mind read about, that this chunk of code makes me want to rewrite that entire section / chapter.

I'm sorry you had to transcribe this bit, thank you. :)

Reply:
Actually, as a green hand, it at least taught me that redis can also be used to solve such problems. How about writing the second edition of Redis In Action? You can teach us more modern and practical methods :P

Reply:
You can do all sorts of crazy things with Redis.

But the publisher owns the trademark for "Redis in Action". I'd like to write another version, I have another 3-5 chapters to write... but it's basically in their court now.